sdks/java/core/src/main/java/org/apache/beam/sdk/util/BufferedElementCountingOutputStream.java

```diff
@@ -65,6 +67,9 @@
  */
 @NotThreadSafe
 public class BufferedElementCountingOutputStream extends OutputStream {
+  private static final int MAX_POOLED = 12;
+  @VisibleForTesting static final ArrayBlockingQueue<ByteBuffer> BUFFER_POOL =
```

Comment:
style/practice: instead of making this `@VisibleForTesting`, make a constructor that is visible for testing that accepts the pool as an argument, and have actual use only go through a static factory method that provides this static pool

Reply:
That would make the buffer pool non-final and have subtle issues running multithreaded tests, which would require concurrency protections that would impact regular performance.

Based on the other comments, it's not necessary to change the queue in use, we can stimulate empty queue and full queue situations easily and test boundary conditions.

Reply:
SGTM but FWIW I did mean everything still final. This:

```java

public class BufferedElementCountingOutputStream extends OutputStream {
  private static final int MAX_POOLED = 12;
  private static final BlockingQueue<ByteBuffer> DEFAULT_BUFFER_POOL =
      new ArrayBlockingQueue<>(MAX_POOLED);

  private final BlockingQueue<ByteBuffer> bufferPool;

  private BufferedElementCountingOutputStream(BlockingQueue<ByteBuffer> bufferPool) {
    this.bufferPool = bufferPool;
  }

  public static BufferedElementCountingOutputStream withDefaultBufferPool() {
    return new BufferedElementCountingOutputStream(DEFAULT_BUFFER_POOL);
  }

  @VisibleForTesting
  static BufferedElementCountingOutputStream withBufferPool(BlockingQueue<ByteBuffer> bufferPool) {
    return new BufferedElementCountingOutputStream(bufferPool);
  }
}
```

... and typing the class that many times practically gave me RSI.